### PR TITLE
Add the option to configure memory ballast for Loki

### LINKED ISSUE #781: Improve cpu/time performance by reusing request goroutines.
During benchmark I realized that 35% of cpu time is spent increasing goroutines stack size see flamegraph below. The remaining 65% is gzip decompression.

image

This is happening because Loki query code is very deep in term of function calls, and because each request creates a new goroutine with an initial stacksize of 2KB, which means each goroutines has to growth by factor 2 (this is how go runtime works) then copy the previous stack in the new one. Once a request is done, the goroutine is thrown away and next request will create a new one, that will also have to grow

> A more detailed explanation can be found here : https://eng.uber.com/optimizing-m3/

My suggestion here is to create a worker/goroutine pool to use for each query request so that we can recycle goroutines instead of throwing them away.

WDYT ? This seems to increase the speed by 35%. I already get rid of all possible recursions we used to have.

/cc @gouthamve @tomwilkie

### PR DESCRIPTION
**Pull request by SasSwart**
**What this PR does / why we need it**:
Adds the option to configure a memory ballast in order to enable reducing garbage collection when needed.
**Which issue(s) this PR fixes**:
Fixes #781

**Special notes for your reviewer**:
PR Based on reading [this](https://blog.twitch.tv/en/2019/04/10/go-memory-ballast-how-i-learnt-to-stop-worrying-and-love-the-heap/), and a comparable [PR in cortex](https://github.com/cortexproject/cortex/pull/1489/files).
**Checklist**
- [x] Documentation added
- [ ] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Review by dannykopping — APPROVED** [review]
LGTM, thanks for this @SasSwart!

**Comment by SasSwart** [comment]
> LGTM, thanks for this @SasSwart! We'll definitely need some documentation about this, though. Do you mind adding that?

On it.

**Comment by SasSwart** [comment]
@dannykopping Done. Added more clarity to flag description as well.

**Comment by dannykopping** [comment]
> @dannykopping Done. Added more clarity to flag description as well.

Sorry, I should've been more clear. We have our docs at https://grafana.com/docs/loki/latest/configuration/ - which is linked to the [`docs`](https://github.com/grafana/loki/tree/main/docs) directory. The comments are worthwhile, for sure, so you can leave those as they are; I think we'll need an update to the [`yaml`](https://grafana.com/docs/loki/latest/configuration/#supported-contents-and-default-values-of-lokiyaml) config overview, perhaps plus a page in [`Operations`](https://grafana.com/docs/loki/latest/operations/) about how/when to use this ballast setting.

This is asking a lot so don't feel obligated; we can add that after merging the PR.

**Comment by cyriltovena** [comment]
I agree I don't know when I should use it either

**Comment by SasSwart** [comment]
Not sure how I missed that. Sorry y'all.

**Comment by SasSwart** [comment]
While adding documentation, I noticed that I wasn't providing a tag for unmarshalling BallastBytes from the config file.

I've moved the BallastBytes value to the root config struct instead of the ConfigWrapper and provided the tag.

I also removed the option to set this as a Command Line argument in favor of the config file.
This seems to align better with the other configuration options.

**Comment by dannykopping** [comment]
> While adding documentation, I noticed that I wasn't providing a tag for unmarshalling BallastBytes from the config file.
> 
> I've moved the BallastBytes value to the root config struct instead of the ConfigWrapper and provided the tag.
> 
> I also removed the option to set this as a Command Line argument in favor of the config file. This seems to align better with the other configuration options.

Thanks for this.

We generally allow for the setting of important config options via both YAML and CLI; would you mind restoring this one? It can be defined in `pkg/loki/loki.go` under `func (c *Config) RegisterFlags(f *flag.FlagSet)` like we do for `auth_enabled`, for example.

**Review by cyriltovena — APPROVED** [review]
LGTM

**Comment by dannykopping** [comment]
Thanks @SasSwart :raised_hands: 
We'll wait for @KMiller-Grafana to approve the docs changes and then merge.

**Comment by bboreham** [comment]
> reduce the amount of time spent resizing goroutine stack sizes.

May I ask what that relates to?  Normally I would say, as the code does, it's to reduce garbage-collection.

**Comment by SasSwart** [comment]
Hey @bboreham 

That would be a wording mistake on my side. When opening the PR, I must've referenced the wrong section of [this issue](https://github.com/grafana/loki/issues/781). Thanks for spotting this.

I seem to have referred to [the top level comment](https://github.com/grafana/loki/issues/781#issue-470003524), whereas my contribution is more related to [this comment](https://github.com/grafana/loki/issues/781#issuecomment-527666824).

I'll update the PR description.

**Review by owen-d — APPROVED** [review]

File: docs/sources/configuration/_index.md
Comment (KMiller-Grafana) [inline docs/sources/configuration/_index.md:96]: Wording suggestion:
Change
> The amount of virtual memory to reserve as a ballast in order to optimise
to be
> The amount of virtual memory in bytes to reserve as ballast in order to optimize

File: docs/sources/operations/scalability.md
Comment (KMiller-Grafana) [inline docs/sources/operations/scalability.md:29]: We use sentence case for headers.  Here's my rewrite of this section of prose:

## Memory ballast

In compute-constrained environments, garbage collection can become a significant performance factor. Frequently-run garbage collection interferes with running the application by using CPU resources. The use of memory ballast can mitigate the issue. Memory ballast allocates extra, but unused virtual memory in order to inflate the quantity of live heap space. Garbage collection is triggered by the growth of heap space usage. The inflated quantity of heap space reduces the perceived growth, so garbage collection occurs less frequently.  

Configure memory ballast using the `ballast_bytes` configuration option.

**Review by KMiller-Grafana — COMMENTED** [review]
This PR was merged in before my doc suggestions were written.  Please consider using the suggestions in further revision.

Memory ballast is cool!

File: docs/sources/operations/scalability.md
Comment (SasSwart) [reply docs/sources/operations/scalability.md:29]: Thanks for the notes @KMiller-Grafana,

I'll prepare a follow up PR with these changes.

**Pull request merged**
